packages/compass-crud/src/components/table-view/cell-renderer.tsx

```diff
@@ -81,182 +90,108 @@ const decrypdedIconStyles = css({
+  useEffect(() => {
+    if (!isEmpty && element) {
+      element.on(Element.Events.Added, handleElementEvent);
+      element.on(Element.Events.Converted, handleElementEvent);
+      element.on(Element.Events.Edited, handleElementEvent);
+      element.on(Element.Events.Reverted, handleElementEvent);
+
+      return () => {
+        element.removeListener(Element.Events.Added, handleElementEvent);
+        element.removeListener(Element.Events.Converted, handleElementEvent);
+        element.removeListener(Element.Events.Edited, handleElementEvent);
+        element.removeListener(Element.Events.Reverted, handleElementEvent);
+      };
     }
-  }
+  }, [isEmpty, element, handleElementEvent]);
 
-  componentDidMount() {
-    if (!this.isEmpty) {
-      this.subscribeElementEvents();
+  const elementLength = useMemo((): number | undefined => {
+    if (!element) {
+      return undefined;
     }
-  }
 
-  componentWillUnmount() {
-    if (!this.isEmpty) {
-      this.unsubscribeElementEvents();
+    if (element.currentType === 'Object') {
+      return Object.keys(element.generateObject() as object).length;
     }
-  }
-
-  subscribeElementEvents() {
-    this.element.on(Element.Events.Added, this.handleElementEvent);
-    this.element.on(Element.Events.Converted, this.handleElementEvent);
-    this.element.on(Element.Events.Edited, this.handleElementEvent);
-    this.element.on(Element.Events.Reverted, this.handleElementEvent);
-  }
-
-  unsubscribeElementEvents() {
-    this.element.removeListener(Element.Events.Added, this.handleElementEvent);
-    this.element.removeListener(
-      Element.Events.Converted,
-      this.handleElementEvent
-    );
-    this.element.removeListener(Element.Events.Edited, this.handleElementEvent);
-    this.element.removeListener(
-      Element.Events.Reverted,
-      this.handleElementEvent
-    );
-  }
-
-  handleElementEvent = () => {
-    this.forceUpdate();
-  };
-
-  handleUndo = (event: React.MouseEvent) => {
-    event.stopPropagation();
-    const oid = this.props.node.data.hadronDocument.getStringId();
-    if (this.element.isAdded()) {
-      this.isDeleted = true;
-      const isArray =
-        !this.element.parent?.isRoot() &&
-        this.element.parent?.currentType === 'Array';
-      this.props.elementRemoved(String(this.element.currentKey), oid, isArray);
-    } else if (this.element.isRemoved()) {
-      this.props.elementAdded(
-        String(this.element.currentKey),
-        this.element.currentType,
-        oid
-      );
-    } else {
-      this.props.elementTypeChanged(
-        String(this.element.currentKey),
-        this.element.type,
-        oid
-      );
+    if (element.currentType === 'Array' && element.elements) {
+      return element.elements.size;
     }
-    this.element.revert();
-  };
-
-  handleDrillDown(event: React.MouseEvent) {
-    event.stopPropagation();
-    this.props.drillDown(this.props.node.data.hadronDocument, this.element);
-  }
+  }, [element]);
```

Comment:
This memo on element here (and probably everywhere else where you have this as a dep) is an issue: element reference itself is stable, and so even though you have the subscription to the element change events that call forceUpdate, the values are still kept intact by the memo logic. As a repro you can try editing an array and then reverting the changes, you can observe that the length stays the same on revert (and the same works currently on main):

![Kapture 2025-06-17 at 17 44 52](https://github.com/user-attachments/assets/f0b218ec-e4cf-4263-b744-fb55c3be511c)